From 3a75d1db61fd420da27f27d345c5d34c53626be7 Mon Sep 17 00:00:00 2001
From: John Howard <howardjohn@google.com>
Date: Thu, 17 Mar 2022 18:17:05 -0700
Subject: pilot: add feature flag to opt out of ALPN filter (#37950)

Useful when using envoyproxy/envoy builds without Istio filters compiled
in
---
 pilot/pkg/features/pilot.go                    | 4 ++++
 pilot/pkg/networking/core/v1alpha3/listener.go | 6 ++++--
 2 files changed, 8 insertions(+), 2 deletions(-)

diff --git a/pilot/pkg/features/pilot.go b/pilot/pkg/features/pilot.go
index aa42fa127d..17bcdb5e22 100644
--- a/pilot/pkg/features/pilot.go
+++ b/pilot/pkg/features/pilot.go
@@ -475,6 +475,10 @@
 		"If true, pilot will add metadata exchange filters, which will be consumed by telemetry filter.",
 	).Get()
 
+	ALPNFilter = env.RegisterBoolVar("PILOT_ENABLE_ALPN_FILTER", true,
+		"If true, pilot will add Istio ALPN filters, required for proper protocol sniffing.",
+	).Get()
+
 	WorkloadEntryAutoRegistration = env.RegisterBoolVar("PILOT_ENABLE_WORKLOAD_ENTRY_AUTOREGISTRATION", true,
 		"Enables auto-registering WorkloadEntries based on associated WorkloadGroups upon XDS connection by the workload.").Get()
 
diff --git a/pilot/pkg/networking/core/v1alpha3/listener.go b/pilot/pkg/networking/core/v1alpha3/listener.go
index 0e0e595533..abd6b428eb 100644
--- a/pilot/pkg/networking/core/v1alpha3/listener.go
+++ b/pilot/pkg/networking/core/v1alpha3/listener.go
@@ -1457,8 +1457,10 @@ func buildHTTPConnectionManager(listenerOpts buildListenerOpts, httpOpts *httpLi
 	}
 
 	// append ALPN HTTP filter in HTTP connection manager for outbound listener only.
-	if listenerOpts.class != istionetworking.ListenerClassSidecarInbound {
-		filters = append(filters, xdsfilters.Alpn)
+	if features.ALPNFilter {
+		if listenerOpts.class != istionetworking.ListenerClassSidecarInbound {
+			filters = append(filters, xdsfilters.Alpn)
+		}
 	}
 
 	// TypedPerFilterConfig in route needs these filters.
-- 
2.35.3

